### PR TITLE
Implement HyperNova (Cyclefold) Folding

### PR DESCRIPTION
**Pull request by sjudson**
Working towards: https://github.com/nexus-xyz/nexus-zkvm/issues/76

The HyperNova folding is implemented in a manner for use as a subprotocol in `/src/folding/hypernova/nimfs.rs`, and then `/src/folding/hypernova/cyclefold/nimfs.rs` contains the cyclefold implementation using it as a subroutine.

### CODE REVIEW

**Comment by dorebell** [comment]
it looks like this is generating some compiler warnings that are preventing CI from passing - you should fix these.

**Comment by slumber** [comment]
@sjudson just add `#![allow(unused)]` at the top of each module, we had those around before finalizing IVC and having modules public.

**Review by dorebell — APPROVED** [review]
nice work! i don't think any of my notes affect correctness, so i think you can respond to them to your liking and then go ahead and merge. 

the one substantial thing is that i think using hypernova for the secondary instance as well as the primary will potentially result in better overall performance. the cost of committing to witnesses for the secondary circuit is quite small compared to the primary, but fewer group operations for verifying the secondary folds in the primary circuit might make the primary circuit smaller. but i'm not certain one way or another. in any case, this change could go in a separate PR if we decide to do it.

File: nova/src/folding/hypernova/nimfs.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
+#![allow(unused)]
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/folding/hypernova/nimfs.rs:1]: There's a name clash as hypernova paper also calls the scheme "multifolding", because there're 2 different LCCS and CCS relations. I was thinking of renaming cyclefold to something else during the last refactoring, but left it as is.

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+            shape_secondary,
+            pp_secondary,
+            U_secondary,
+            W_secondary,
+            &W_comm_trace.0,
+            &W_comm_trace.1,
+        )?;
+        random_oracle.absorb_non_native(&W_comm_trace.0);
+        random_oracle.absorb(&commitment_T.into_affine());
+        random_oracle.absorb(&cast_field_element_unique::<G1::BaseField, G1::ScalarField>(&rho));
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:115]: I disagree:

0. CycleFold paper describes the folding scheme with R1CS for secondary instance, so @sjudson implementation is accurate in this regard.
1. There's only one secondary instance to be folded - proof for `commitment_W`. (in fact, folding multiple instances with cyclefold doesn't add much overhead -- the main contributor is non-native multiplication and we only reduce `X` once at the end)
2. Turning `RelaxedR1CSInstance` into `LCCSInstance` is going to increase ~11 non-native absorbtions to 11 + 4 + 3 (with R1CS, I think?), and this is the main constraints contributor. 
3. I'm not really sure about gadgets for non-native folding off the top of my head, it looks like it's going to add a huge overhead.

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+            commitment_T,
+            commitment_W_proof,
+            hypernova_proof,
+            _poly_commitment: PhantomData,
+        };
+
+        Ok((proof, (folded_U, folded_W), (U_secondary, W_secondary)))
+    }
+
+    #[cfg(any(test, feature = "spartan"))]
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:135]: This is probably a copy-paste.

I suggest we vote whether we want to feature-gate anything. `spartan` is defaulted everywhere in the workspace, and disabling it will break hypernova because of traits and zeromorph dependencies. I proposed adding `feature = "spartan"` as I thought people might want to opt out from it (microsoft Nova e.g. doesn't care).
So either we handle it properly or drop it for the time being.

@sjudson @dorebell wdyt?

(this PR can do whatever compiles **and** correct w.r.t. hypernova paper, so we don't block it from merging with this discussion)

File: nova/src/absorb.rs [hunk 4]
```diff
@@ -124,6 +126,60 @@ where
     }
 }
 
+/// See the above comment for [`R1CSInstance`] non-native absorb implementation.
```
Comment (slumber) [inline nova/src/absorb.rs:129]: I don't think you need these implementations at all

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+        // The PolyCommitment trait only guarantees the commitment can be represented by a vector of field elements. However,
+        // it makes sense to implement HyperNova Cyclefold assuming the commitment is just a single field element, because we
+        // will be using such a scheme for now (Zeromorph) and it leads to a minimally-sized secondary circuit. We expect this
+        // conversion will panic as unimplemented!() for any incompatible such commitment scheme.
+
+        let s_U_commitment_W = U.commitment_W.clone().into_single();
+        let s_u_commitment_W = u.commitment_W.clone().into_single();
+
+        let g_out = s_U_commitment_W + s_u_commitment_W * rho_scalar;
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:91]: Think so too, with `into_single()`

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    ) -> Result<
+        (
+            Self,
+            (LCCSInstance<G1, C1>, CCSWitness<G1>),
+            (RelaxedR1CSInstance<G2, C2>, RelaxedR1CSWitness<G2>),
+        ),
+        Error,
+    > {
+        let mut random_oracle = RO::new(config);
+
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:69]: You also need it here for secondary instance, so it has to be shared.

File: nova/src/ccs/mle.rs [hunk 7]
```diff
@@ -41,10 +41,44 @@ pub fn vec_to_mle<F: PrimeField>(z: &[F]) -> DenseMultilinearExtension<F> {
...
+
+    let mut z = z.to_owned();
+    z.resize(n.next_power_of_two(), F::zero());
+
+    // Explanation of reversing:
+    //
+    //   At present for polynomial commitments we use the ones from the spartan repo,
+    //   because they work nicely over its multilinear extension representation.
+    //
+    //   This is not true of the ark-poly-commit implementations, which expect for
```
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/ccs/mle.rs:59]: It does implement `Polynomial`, but from what I see the `ark_poly_commit` library implementations are bound to `DenseUVPolynomial` and `DenseMVPolynomial` as appropriate, which `DenseMultilinearExtension` doesn't implement.

File: nova/src/ccs/mod.rs [hunk 8]
```diff
@@ -1,17 +1,20 @@
...
 use ark_ec::{AdditiveGroup, CurveGroup};
-use ark_ff::Field;
+use ark_ff::{Field, PrimeField};
 use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
 use ark_spartan::polycommitments::PolyCommitmentScheme;
+use ark_std::fmt;
 
-use ark_std::Zero;
-use std::ops::Neg;
+use ark_std::{fmt::Display, ops::Neg, Zero};
```
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/ccs/mod.rs:9]: (won't let me directly apply but will incorporate)

File: nova/src/folding/hypernova/nimfs.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/folding/hypernova/nimfs.rs:1]: Yeah I handled this by just alias'ing the import into the cyclefold module to `HNProof`, which works well enough

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:135]: Yeah it's a copy-paste. I don't have a strong opinion about whether to keep it gated: the two of you have a much better sense of the intended evolution of the repo than I do at present.

**Review by dorebell — COMMENTED** [review]
> I disagree:

0. CycleFold paper describes the folding scheme with R1CS for secondary instance, so @sjudson implementation is accurate in this regard.
1. There's only one secondary instance to be folded - proof for `commitment_W`. (in fact, folding multiple instances with cyclefold doesn't add much overhead -- the main contributor is non-native multiplication and we only reduce `X` once at the end)
2. Turning `RelaxedR1CSInstance` into `LCCSInstance` is going to increase ~11 non-native absorbtions to 11 + 4 + 3 (with R1CS, I think?), and this is the main constraints contributor. 
3. I'm not really sure about gadgets for non-native folding off the top of my head, it looks like it's going to add a huge overhead. 

does the cyclefold paper say much about hypernova? i didn't think this combination was presented anywhere in much detail, so idk how much consistency matters here. but yeah, it's plausible to me that the hypernova sumcheck/hashing overhead wouldn't be worth it to save a group operation in the main circuit. 

actually, this is basically the same question as whether or not hypernova makes the main augmented circuit smaller. Does it? (even if the answer is no, hypernova halves the number of witness commitments needed, so probably wins on the balance anyway).

**Review by dorebell — COMMENTED** [review]
> Yeah it's a copy-paste. I don't have a strong opinion about whether to keep it gated: the two of you have a much better sense of the intended evolution of the repo than I do at present. 

i don't have a strong opinion either, but maybe removing feature gates for now is fine - compilation isn't that slow, and the spartan feature doesn't change the behavior of anything, only adds features. at some point when the codebase is in a more stable place, we can think about how best to carve out some functionality into features.

File: nova/src/ccs/mod.rs [hunk 9]
```diff
@@ -229,10 +305,50 @@ impl<G: CurveGroup, C: PolyCommitmentScheme<G>> LCCSInstance<G, C> {
...
+    pub fn fold(
+        &self,
+        U2: &CCSInstance<G, C>,
+        rho: &G::ScalarField,
+        rs: &[G::ScalarField],
+        sigmas: &[G::ScalarField],
+        thetas: &[G::ScalarField],
+    ) -> Result<Self, Error> {
+        let (uX1, comm_W1) = (&self.X, self.commitment_W.clone());
+        let (oX2, comm_W2) = (&U2.X, U2.commitment_W.clone());
```
[the remark this answers is not in the document]
Comment (dorebell) [reply nova/src/ccs/mod.rs:318]: oh i see - in other places, we use the convention that `u` is part of `X`, so maybe we should stick with that. definitely not a big deal tho.

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    G1: SWCurveConfig<BaseField = G2::ScalarField, ScalarField = G2::BaseField>,
+    G2: SWCurveConfig,
+    C1: PolyCommitmentScheme<Projective<G1>>,
+    C2: CommitmentScheme<Projective<G2>>,
+    G1::BaseField: PrimeField + Absorb,
+    G2::BaseField: PrimeField + Absorb,
+    RO: CryptographicSponge,
+{
+    pub fn prove(
+        _ck: &C1::PolyCommitmentKey,
```
Comment (dorebell) [inline nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:48]: why is this here? there shouldn't be any need to compute polynomial commitments anywhere. if we're going to hash the key as part of initializing the random oracle (never a bad idea, but i'm not certain whether it's necessary), we should use the `EvalVerifierKey` instead of the `PolyCommitmentKey` (the latter is huge).

**Comment by dorebell** [comment]
The changes look good, and I think this is ready to merge. Good work! 

We might want to try out using hypernova for the secondary instance sometime, but that's definitely something we can experiment with once we have the rest of this working.

File: nova/src/folding/hypernova/cyclefold/nimfs/mod.rs (hunk 10 above)
Comment (sjudson) [reply nova/src/folding/hypernova/cyclefold/nimfs/mod.rs:48]: Ah sorry, left this in when things were in progress and wasn't sure if I'd need it in some way. Thanks for the catch, and will remove.

File: spartan/src/polycommitments/mod.rs [hunk 11]
```diff
@@ -41,13 +41,23 @@ pub trait PolyCommitmentTrait<G: CurveGroup>:
...
   // this should be the commitment to the zero vector of length n
   fn zero(n: usize) -> Self;
+
+  fn into_affine(self) -> Vec<G::Affine> {
+    self.into().iter().map(|c| c.into_affine()).collect()
+  }
+
+  // for commitments in the form of a single field element (e.g., Zeromorph) this should return it
+  // for commitments unable to be representated as such (e.g., Hyrax) it should panic unimplemented!()
+  fn into_single(self) -> G;
```
Comment (slumber) [inline spartan/src/polycommitments/mod.rs:60]: Wouldn't having just one method
```rust
/// Convert commitment into an affine point.
///
/// This method should only be implemented in case the commitment scheme returns a single curve point.
fn into_affine(self) -> Option<G::Affine> {
    None
}
```
be sufficient?

Looks cleaner to me. We can then panic inside the folding scheme implementation. Returning error doesn't make sense as it cannot be handled as such -- the commitment scheme is not suitable.

File: spartan/src/polycommitments/mod.rs [hunk 12]
```diff
@@ -51,13 +51,9 @@ pub trait PolyCommitmentTrait<G: CurveGroup>:
...
   fn zero(n: usize) -> Self;
 
-  fn into_affine(self) -> Vec<G::Affine> {
-    self.into().iter().map(|c| c.into_affine()).collect()
-  }
-
-  // for commitments in the form of a single field element (e.g., Zeromorph) this should return it
-  // for commitments unable to be representated as such (e.g., Hyrax) it should panic unimplemented!()
-  fn into_single(self) -> G;
+  // for commitments in the form of a single field element (e.g., Zeromorph) this should return Some(x)
```
Comment (slumber) [inline spartan/src/polycommitments/mod.rs:54]: As per my previous comment, you can provide default implementation which returns `None`.

The name is misleading, I still suggest `into_affine` or even `try_into_affine` because of `Option`. We want this method for hashing, and protective coordinates aren't unique. No use for it otherwise.

Finally, consider adding something similar to the suggested doc comment in the code snippet, for example developers will be able to hover over this function and see what it does, instead of how it _should_ be implemented. (Meaning /// as well)

**Review by slumber — APPROVED** [review]
